### PR TITLE
Add `length` kernel support for List Array

### LINKED ISSUE #1470: Add `length` kernel support for `ListArray`
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**
Currently, the `length` kernel only works with `StringArray`. It is reasonable to make it support the `ListArray`.

**Describe the solution you'd like**
it should be possible to call something like `arrow::compute::length(list_array)` and return an array that contains the length of all lists.

Related issue: #1464

### PR DESCRIPTION
**Pull request by HaoYang670**
# Which issue does this PR close?

Closes #1470 .

# What changes are included in this PR?
1. add `length` kernel support for list array
2. do some clean up work, including: rename private functions (make names more consistent), format the generic type bounds and update the doc.

# Are there any user-facing changes?
Doc of the public function `length` is updated.

### CODE REVIEW

File: arrow/src/compute/kernels/length.rs [hunk 1]
```diff
@@ -110,20 +123,24 @@ where
...
 
-/// Returns an array of Int32/Int64 denoting the number of bytes in each value in the array.
+/// Returns an array of Int32/Int64 denoting the length of each value in the array.
+/// For list array, length is the number of elements in each list.
+/// For string array and binary array, length is the number of bytes of each value.
 ///
-/// * this only accepts StringArray/Utf8, LargeString/LargeUtf8, BinaryArray and LargeBinaryArray
+/// * this only accepts ListArray/LargeListArray, StringArray/LargeStringArray and BinaryArray/LargeBinaryArray
 /// * length of null is null.
 /// * length is in number of bytes
```
Comment (viirya) [inline arrow/src/compute/kernels/length.rs:132]:
```suggestion
/// * length of null is null.
```

**Review by viirya — APPROVED** [review]
Looks good. Thanks.

File: arrow/src/compute/kernels/length.rs (hunk 1 above)
Comment (HaoYang670) [reply arrow/src/compute/kernels/length.rs]: Done!

**Review by alamb — COMMENTED** [review]
Thanks @HaoYang670  and @viirya

**Pull request merged**